### ISSUE TITLE
CUDA 9.2

### ISSUE THREAD

**Issue by nakul02** [jcuda/jcuda-main#25]
CUDA 9.2 has been our for about 2 weeks now.
Would it be a good idea to have a maven jcuda release for 9.2?
I could provide binaries for windows, x86_64, ppc64le and possibly Mac.

**Comment by jcuda** [comment]
Sure, there should be an update ASAP (I wonder why I didn't receive a notification about the update via the NVIDIA newsletter...?!). 

The unfortunate thing here is that the update for *9.1* is still pending. I'm currently somewhat short of time, but will increase the priority for the update accordingly.

**Comment by nakul02** [comment]
Is there an easy way to update the version number? or it is a manual process across all the projects?
I can try and create binaries for 9.1 and 9.2.

**Comment by jcuda** [comment]
It may not only be the version number. I'll have to check whether new functions have been introduced or function signatures changed.

**Comment by nakul02** [comment]
[Here](https://send.firefox.com/share/b172eb8f42) are binaries for jcuda 0.9.0d for apple x86_64, in case you wanted to make a release. They do require llvm 8 and don't seem to work with llvm 9.

**Comment by jcuda** [comment]
Thanks for this contribution! I'm not sure whether I'll try to extend the current Maven release with the Mac binaries. I think it *could* be possible, and it could be a nice test case to see whether adding natives in general is possible, but I would have to figure this out first, and currently, I think that the release of 9.1 and 9.2 have higher priority.

If you could contribute the 9.2 binaries, that would be great, but I still have to prepare the 9.2 release.

In the meantime, I have updated to 9.1, and the binaries should be available in Maven Central soon. (I actually wanted to start the update for 9.2 today, but burnt some time with a VirtualBox that refused to start until the NVIDIA drivers had been uninstalled...)

I'll try to do the update for 9.2 in the next days (it should not take nearly as long as the 9.1 one - sorry for the delay here...)

**Comment by nakul02** [comment]
Agreed, the adding of Mac binaries will serve as a useful test case if being able to add binaries for other platforms later. 

I think the 9.2 is probably higher priority than this. But just in case all binaries for 9.2 aren’t available when you decide to make a release, trying out the Mac binaries for 0.9.0d will serve as a useful test. 

Tag me on this thread or any other when you’ve updated the project for 9.1 and 9.2, ideally making a tag for each release, and I will try to provision ppc64le and Mac machines with NVIDIA cards. 

The current CMakeLists.txt doesn’t compile the jcudnn project. Maybe you could change that with the release for 9.1?

**Comment by jcuda** [comment]
I'll drop you a note as soon as the 9.2 binaries can be built. I usually tag the versions that the binaries are built from (for the current release, the tag is `version-0.9.1-RC00` for all libraries).

This still excludes JCudnn: That's a bit difficult. There are two versions involved, and they change sometimes independently. For example, the current release is built for CUDA 9.1.85 and cuDNN 7.0.5 (for CUDA 9.1, and Windows 7). There is another CUDNN 7.0.x which is for CUDA 9.0. So a single (major) cuDNN version may match different CUDA versions, or vice versa. Additionally, cuDNN is still only available for registered developers and not part of "core CUDA". 

However, I've noticed that people have been particularly interested in JCudnn, and would also prefer to have it "better integrated". I think adding it to the CMakeLists and POMs should be OK, because those who build the binaries will (or should) usually *also* build the JCudnn ones. I'll consider this for 9.2, or maybe for the next version.

**Comment by jcuda** [comment]
Another delay here. NVIDIA has messed up the installer again. Until now I could not install CUDA 9.2 on my PC. The threads in the NVIDIA forums are not really helpful in this regard. If it simply doesn't work, there's not much I can do. But maybe I'll try the Win7 version tomorrow or so.

:-1:

**Comment by jcuda** [comment]
Right now, I'm afraid that 9.2 will be the first CUDA version for which no JCuda release will be created. 

The NVIDIA installers are totally broken (for CUDA as well as for the drivers). Threads like https://devtalk.nvidia.com/default/topic/1024655/cuda-setup-and-installation/cuda-9-0-176-installer-fails-every-time/3 (and many others) seem to be related, but running it as admin did not help, disabling the virus checker did not help, neither the network- nor the local installation worked, unpacking the installer did not help, enabling logging created a 15 MB text file without *any* error message, deactivating the VS integration and "Bad GeForce Experience" did not help, deactivating other install options causes the installer to hang up (yes, clicking a *checkbox* causes it to hang up!), and the recommendation that was given in the NVIDIA forum thread (namely, to "reinstall windows") is about as funny as the videos and GIF files that can be found by googling for "torvalds nvidia". 

Closing this one until the people at NVIDIA get their stuff in order.

**Comment by jcuda** [comment]
OK. During another try (without clicking any of the dangerous checkboxes - thou shalt be wary of those!), it *again* did hang with an indeterminate progress bar, saying "Preparing for installation", like it did maybe 20 times before, and like it is reported in dozens of NVIDIA forum threads. 

But while reading these desperate threads, I found the solution: Waiting. After only about half an hour of *seemingly doing nothing*, the installation continued and finished. 

Seriously? 

Whatever they are doing while they are "Preparing for installation", they should use CUDA, maybe they could make it faster (scnr). 

Reopening, hopefully I can finally do the update now.

**Comment by jcuda** [comment]
Just a short heads up: Most of the updates for CUDA 9.2 have just been committed. There is some new functionality in CUBLAS and cuDNN that is not yet supported (several flavors of callbacks - they can be tricky to map into the Java world sensibly). If someone is eager, he might give the current state a try. 

Otherwise, I'd tag the current state as 0.9.2-RC00 ASAP (i.e. in the next few days), and see whether everything works as expected when compiling the natives in the Linux VM. If this also works well, the other natives could be built as well based on the tagged state.

**Comment by jcuda** [comment]
The `version-0.9.2-RC00` tag has been created, but I have not yet built the binaries on Linux from this state. (I might be able to do this on Thursday). For those who want to provide other binaries (and maybe the Linux ones, if they are quicker than me) : Note that JCudnn has now been added to the build process. (The READMEs have to be updated accordingly).

So the binaries are supposed to be built for CUDA 9.2 and cuDNN 7.1. When the unit tests are run in Maven, the cuDNN library (which is not part of core CUDA) will have to be in a visible path.

**Comment by nakul02** [comment]
Here is linux x86_64 - https://send.firefox.com/download/ba94b24360/#hYA0WdMYQQ4FNVXLd_0L2A

**Comment by jcuda** [comment]
Awesome @nakul02 ! (And impressive - I've still been downloading and installing 9.2 on my VM...). Then I could probably do the release on Thursday, with all binaries that are available until then.

**Comment by nakul02** [comment]
👍 I will try

**Comment by nakul02** [comment]
Here is the ppc64le - https://send.firefox.com/download/ef0fcf5e35/#ayHVzXgSoc2Dhw6oB7735w

**Comment by nakul02** [comment]
Also, Here are the binaries for **JCUDA 9.0** - https://send.firefox.com/download/308ba7404f/#BPPfW3Pbyczf5mfrAw7Rdw in the case you were going to make a release for ppc64le and mac for version 0.9.0d.

**Comment by nakul02** [comment]
And binaries for JCUDA 9.1 ppc64le - 
https://send.firefox.com/download/f7d3667831/#RGIwcPE4nWj9FPc3HZMnIA

**Comment by jcuda** [comment]
Thanks @nakul02 but I have to apologize: I obviously should have downloaded the files immediately - the links are expired (I'm not sure what the expiration period is, though - maybe you can attach the files directly here, or send them via mail to jcuda `@` jcuda.org ?

**Comment by nakul02** [comment]
Sent it to you through email.

**Comment by jcuda** [comment]
A huge **thanks!** to @nakul02 for providing the Linux binaries so quickly. 

Version 0.9.2 has just been released and should be available in Maven Central soon, at the usual coordinates: 

    <dependency>
        <groupId>org.jcuda</groupId>
        <artifactId>jcuda</artifactId>
        <version>0.9.2</version>
    </dependency>
    <dependency>
        <groupId>org.jcuda</groupId>
        <artifactId>jcublas</artifactId>
        <version>0.9.2</version>
    </dependency>
    <dependency>
        <groupId>org.jcuda</groupId>
        <artifactId>jcufft</artifactId>
        <version>0.9.2</version>
    </dependency>
    <dependency>
        <groupId>org.jcuda</groupId>
        <artifactId>jcusparse</artifactId>
        <version>0.9.2</version>
    </dependency>
    <dependency>
        <groupId>org.jcuda</groupId>
        <artifactId>jcusolver</artifactId>
        <version>0.9.2</version>
    </dependency>
    <dependency>
        <groupId>org.jcuda</groupId>
        <artifactId>jcurand</artifactId>
        <version>0.9.2</version>
    </dependency>
    <dependency>
        <groupId>org.jcuda</groupId>
        <artifactId>jnvgraph</artifactId>
        <version>0.9.2</version>
    </dependency>
    <dependency>
        <groupId>org.jcuda</groupId>
        <artifactId>jcudnn</artifactId>
        <version>0.9.2</version>
    </dependency>

Let me know if you encounter any difficulties.

---

<sub>I'll leave this issue open until the Website and READMEs have been updated</sub>

**Comment by nakul02** [comment]
Here are the mac binaries for 0.9.2 - https://send.firefox.com/download/fe595e3660/#ueII8OYB7HTB569y7qEeCg
(sorry i couldn't send this in earlier)

**Comment by jcuda** [comment]
So then I'd rather try to add these to maven (and if it works, also the other ones that you provided). I won't have toooo much time to experiment around there, but will simply give it a try.

**Comment by jcuda** [comment]
So after a first try, it does not work directly.

In order to bring the libraries into Maven Central, they are uploaded as "bundles" at https://oss.sonatype.org . Such a bundle basically contains the POM+JAR+Sources+JavaDoc, all signed. If the bundle does not contain these elements, the upload is rejected. 

The problem now is that uploading another bundle, say, for `jcuda-natives`, which now should only contain the MacOS natives, is not possible, because the artifact `jcuda-natives-0.9.2` has already been released. 

If anybody has an idea of how to solve this, I'd like to try it out. Otherwise, I'd have to ask, maybe at Stack Overflow or directly at Sonatype whether this is possible at all. 

---

@nakul02 BTW: The Mac binaries JARs contained the 0.9.2 **and** the 0.9.0d version of the native library. I'm basically aware of this problem, but have to see whether I can delete the unwanted libraries during a `mvn clean`, or filter the libraries to be included to *only* contain the one for the current version. I've opened https://github.com/jcuda/jcuda-main/issues/26 for that - it should not be so hard, but I will have to see when I have the chance to tackle this.
